src/folding/nova/circuits.rs

```diff
@@ -0,0 +1,275 @@
+
+        let ci = CommittedInstance::<Projective> {
+            cmE: Projective::rand(&mut rng),
+            u: Fr::rand(&mut rng),
+            cmW: Projective::rand(&mut rng),
+            x: vec![Fr::rand(&mut rng); 1],
+        };
+
+        let cs = ConstraintSystem::<Fr>::new_ref();
+        let ciVar =
+            CommittedInstanceE1Var::<Projective>::new_witness(cs.clone(), || Ok(ci.clone()))
+                .unwrap();
+        assert_eq!(ciVar.u.value().unwrap(), ci.u);
+        assert_eq!(ciVar.x.value().unwrap(), ci.x);
+        dbg!(cs.num_constraints());
+
+        // check the instantiation of the CycleFold side:
+        let cs = ConstraintSystem::<Fq>::new_ref();
+        let ciVar =
+            CommittedInstanceE2Var::<Projective, GVar>::new_witness(cs.clone(), || Ok(ci.clone()))
+                .unwrap();
+        assert_eq!(ciVar.cmE.value().unwrap(), ci.cmE);
+        assert_eq!(ciVar.cmW.value().unwrap(), ci.cmW);
+        dbg!(cs.num_constraints());
+    }
+
+    #[test]
+    fn test_nifs_gadget() {
+        let r1cs = get_test_r1cs();
+        let z1 = get_test_z(3);
+        let z2 = get_test_z(4);
+        let (w1, x1) = r1cs.split_z(&z1);
+        let (w2, x2) = r1cs.split_z(&z2);
+
+        let w1 = Witness::<Projective>::new(w1.clone(), r1cs.A.n_rows);
+        let w2 = Witness::<Projective>::new(w2.clone(), r1cs.A.n_rows);
+
+        let mut rng = ark_std::test_rng();
+        let pedersen_params = Pedersen::<Projective>::new_params(&mut rng, r1cs.A.n_cols);
+
+        // compute committed instances
+        let ci1 = w1.commit(&pedersen_params, x1.clone());
+        let ci2 = w2.commit(&pedersen_params, x2.clone());
+
+        // get challenge from transcript
+        let config = poseidon_test_config::<Fr>();
+        let mut tr = PoseidonTranscript::<Projective>::new(&config);
+        let r_bits = tr.get_challenge_nbits(128);
+        let r_Fr = Fr::from_bigint(BigInteger::from_bits_le(&r_bits)).unwrap();
+
+        let (_w3, ci3, _T, cmT) =
+            NIFS::<Projective>::prove(&pedersen_params, r_Fr, &r1cs, &w1, &ci1, &w2, &ci2);
+
+        let cs = ConstraintSystem::<Fr>::new_ref();
+
+        let rVar = FpVar::<Fr>::new_witness(cs.clone(), || Ok(r_Fr)).unwrap();
+        let ci1Var =
+            CommittedInstanceE1Var::<Projective>::new_witness(cs.clone(), || Ok(ci1.clone()))
+                .unwrap();
+        let ci2Var =
+            CommittedInstanceE1Var::<Projective>::new_witness(cs.clone(), || Ok(ci2.clone()))
+                .unwrap();
+        let ci3Var =
+            CommittedInstanceE1Var::<Projective>::new_witness(cs.clone(), || Ok(ci3.clone()))
+                .unwrap();
+
+        NIFSGadget::<Projective>::verify(
+            rVar.clone(),
+            ci1Var.clone(),
+            ci2Var.clone(),
+            ci3Var.clone(),
+        )
+        .unwrap();
+        assert!(cs.is_satisfied().unwrap());
+        // dbg!(cs.num_constraints());
```

Comment:
Leftover?

Reply:
This was because in the future is useful to 're-activate' the print to see how many constraints it takes. But agree that is not ideal to have things commented out.
Maybe I should just leave it without being commented out, since by default will not be printed?

Reply:
I'd remove it TBH. But again, we can leave it for now for further testing if you think it might be needed.

Reply:
Ok, removed all the `dbg!(cs.num_constraints());`.